vir/src/callable_idents.rs

```diff
@@ -0,0 +1,140 @@
+use crate::{ExprGen, PredicateAppGen, Type, PredicateAppGenData, StmtGenData, MethodCallGenData, VirCtxt};
+use private::*;
+
+#[derive(Debug, Clone, Copy)]
+pub struct FunctionIdent<'vir, A: Arity>(&'vir str, A);
+impl<'vir, T: Arity> FunctionIdent<'vir, T> {
+    pub const fn new(name: &'vir str, args: T) -> Self {
+        Self(name, args)
+    }
+    pub fn name(&self) -> &'vir str {
+        self.0
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct MethodIdent<'vir, A: Arity>(&'vir str, A);
+impl<'vir, T: Arity> MethodIdent<'vir, T> {
+    pub const fn new(name: &'vir str, args: T) -> Self {
+        Self(name, args)
+    }
+    pub fn name(&self) -> &'vir str {
+        self.0
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct PredicateIdent<'vir, A: Arity>(&'vir str, A);
+impl<'vir, T: Arity> PredicateIdent<'vir, T> {
+    pub const fn new(name: &'vir str, args: T) -> Self {
+        Self(name, args)
+    }
+    pub fn name(&self) -> &'vir str {
+        self.0
+    }
+}
+
+/// Module to prevent anyone else from implementing Arity
+mod private {
+    use super::*;
+    pub trait Arity {}
+    impl<const N: usize> Arity for KnownArity<'_, N> {}
+    impl Arity for UnknownArity<'_> {}
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct KnownArity<'vir, const N: usize>([Type<'vir>; N]);
+impl<'vir, const N: usize> KnownArity<'vir, N> {
+    pub const fn new(types: [Type<'vir>; N]) -> Self {
+        Self(types)
+    }
+}
+pub type NullaryArity = KnownArity<'static, 0>;
+pub type UnaryArity<'vir> = KnownArity<'vir, 1>;
+pub type BinaryArity<'vir> = KnownArity<'vir, 2>;
+
+#[derive(Debug, Clone, Copy)]
+pub struct UnknownArity<'vir>(&'vir [Type<'vir>]);
+impl<'vir> UnknownArity<'vir> {
+    pub const fn new(types: &'vir [Type<'vir>]) -> Self {
+        Self(types)
+    }
+}
+
+// Func arity known at compile time
+
+// TODO: deduplicate all the similar code across the `apply` methods
```

Comment:
What is there to deduplicate? The methods seem pretty lightweight.

Reply:
I imagine that at some point we'll be checking that the `args` provided have the same type as the formal args as well as storing the result type and putting that in the returned value, so the functions might grow in size. The duplication that I'm complaining about is mainly that each `...Ident` has to implement the same `&[T]` thing twice and we could separate out the length check and the construction of the result from `&[T]` parts.

Reply:
Changed to reduce the expected duplication and removed the comment